onmt/inference.py

```diff
@@ -13,21 +13,121 @@
 opt = OPT
 opt.gpu = -1 # Use CPU instead of defaulting to GPU 0
 
-model = Img2LatexModel(name=model_path, opt=opt)
+model = Img2LatexModel( name=model_path,opt=opt)
 
-def predict_file(file_name):
+def _predict_with_batch(fns, batch_size):
+    """
+
+    :param fns: results from cv2.imread (numpy array)
+    :param batch_size: number of samples to be processed
+    :return: predicts, confidents
+    """
+    assert batch_size >= 1
+
+    samples_len = len(fns)
+    batches = [(s,min(s + batch_size, samples_len)) for s in range(0, samples_len, batch_size)]
+
+    for (s,e) in batches:
+        _fn_inps = fns[s:e]
+        try:
+            _inps = [cv2.imread(__fn_inps, 0) for __fn_inps in _fn_inps]
+
+            _predicts, _confidents = model.predict(_inps, batch_size=batch_size)
+
+            yield (_predicts, _confidents)
+        except:
+            # if meet the exception, try to run with batch size 1 instead
+            _predicts, _confidents = [], []
+
+            for _fn_inp in _fn_inps:
+                try:
+                    _inp = [cv2.imread(_fn_inp)]
+                    _predict, _confident = model.predict(_inp, batch_size=1)
+
+                    _predicts += _predict
+                    _confidents += _confident
+                except:
+                    _predicts += [None]
+                    _confidents += [None]
+
+            yield (_predicts, _confidents)
+
+
+def predict_file(file_name, batch_size=1):
     global model
 
-    results = {}
     try:
-        predict, confident = model.predict([cv2.imread(file_name, 0)])
+        fns, results = [], []
+        if type(file_name) is str:
+            batch_size = 1
+            fns = [file_name]
+        elif type(file_name) is list:
+            fns = file_name
+        else:
+            raise Exception("file_name parameter must be string or colection of string ...")
 
-        results["prediction"] = predict[-1]
-        results["confidence"] = confident[-1]
+        for (predicts, confidents) in _predict_with_batch(fns=fns, batch_size=batch_size):
+            for (_predict, _confident) in zip(predicts, confidents):
+                if _predict is None or _confident is None:
+                    _dct = {
+                        "success": False
+                    }
+                else:
+                    _dct = {
+                        "prediction" : _predict,
+                        "confidence" : _confident,
+                        "success" : True
+                    }
 
-        results["success"] = True
+                results += [_dct]
+
+        if len(results) == 1: results = results[0] # for compatible with the previous committed code.
```

Comment:
we should check against batch_size == 1 to avoid possible divergence in between the results len and the initial input batch_size. If for some reasons results len is 1 even though batch_size >= 1, this can introduce bugs
```suggestion
        if batch_size == 1: results = results[0] # for compatible with the previous committed code.
```

Reply:
The main purpose of mine when writing the above code is to make the newer code more compatible with the old one. In the previous, we feed only one file name, and output a dictionary (not list of dict). The above code is for only very special case when results is a list of just one element (mainly because we pass just one file name at that time). If we check batch_size == 1: then flatten (results == results[0]), so what happened if the results is a list with more than one elemetns ?, For that reason,  I will leave that snippet code above as it is.

Reply:
No, in my point of view it's right, because i tried to flatten the list if it has only element (because **the previous version we return the dictionary, not list of dictionary**.) no matter if the input contains one file name or batch size is equal to 1 .